feat(swarm): allow behaviours to share addresses of peers (#4371)
paritytech/rust-libp2p: swarm/CHANGELOG.md

> ## 0.44.2 - unreleased

Reviewer: ```suggestion
## 0.44.2
```

> - Add `FromSwarm::NewExternalAddrOfPeer` event.
> - Add `libp2p::swarm::add_peer_address` that broadcasts address through `FromSwarm::NewExternalAddrOfPeer`.
> - Add `libp2p_swarm::behaviour::PeerAddresses` which is used as a cache for storing known addresses
>   of connected peers.
>   See [PR 4371](https://github.com/libp2p/rust-libp2p/pull/4371)

Reviewer: ```suggestion
- Allow `NetworkBehaviour`s to share addresses of peers.
  This is enabled via the new `ToSwarm::NewExternalAddrOfPeer` event.
  The address is broadcast to all behaviours via `FromSwarm::NewExternalAddrOfPeer`.
  Protocols that want to collect these addresses can use the new `PeerAddresses` utility.
  See [PR 4371](https://github.com/libp2p/rust-libp2p/pull/4371).
```

## 0.44.2 - unreleased

- Add `FromSwarm::NewExternalAddrOfPeer` event.

>   See [PR 4371](https://github.com/libp2p/rust-libp2p/pull/4371)

Reviewer: Let's also document the new `Swarm` API please. That is what users will interact with most.

Author: Added more info, but I'm not sure if it's sufficient for someone using the library :thinking:

> ## 0.44.0

Reviewer: ```suggestion
## 0.44.0 - unreleased
```

Author: :heavy_check_mark:
